lib/Model/Pass/FixModel.cpp

```diff
@@ -20,86 +20,83 @@ static Logger<> ModelFixLogger("model-fix");
+  }
 
-    // Filter out empty arrays.
-    bool FoundEmptyArray = false;
-    for (const model::QualifiedType &QT : T->edges()) {
-      auto IsEmptyArray = [](const model::Qualifier &Q) {
-        return Q.Kind() == model::QualifierKind::Array && Q.Size() == 0;
-      };
-      auto Iterator = llvm::find_if(QT.Qualifiers(), IsEmptyArray);
-      if (Iterator != QT.Qualifiers().end()) {
-        ToDrop.insert(T.get());
-        FoundEmptyArray = true;
-      }
-    }
-    if (FoundEmptyArray)
-      continue;
-
-    // Filter out invalid PrimitiveTypes.
-    auto *ThePrimitiveType = dyn_cast<PrimitiveType>(T.get());
-    if (ThePrimitiveType) {
-      if (ThePrimitiveType->PrimitiveKind() == PrimitiveTypeKind::Invalid)
-        ToDrop.insert(T.get());
-    }
+  // Filter out empty arrays.
+  for (const model::QualifiedType &QT : T->edges()) {
+    auto IsEmptyArray = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Array && Q.Size() == 0;
+    };
+    auto Iterator = llvm::find_if(QT.Qualifiers(), IsEmptyArray);
+    if (Iterator != QT.Qualifiers().end())
+      return true;
+  }
+
+  // Filter out invalid PrimitiveTypes.
+  auto *ThePrimitiveType = dyn_cast<PrimitiveType>(T.get());
+  if (ThePrimitiveType) {
+    if (ThePrimitiveType->PrimitiveKind() == PrimitiveTypeKind::Invalid)
+      return true;
+  }
+
+  // Filter out invalid functions.
+  auto *FunctionType = dyn_cast<CABIFunctionType>(T.get());
+  if (FunctionType) {
+    // Remove functions with more than one `void` argument.
+    for (auto &Group : llvm::enumerate(FunctionType->Arguments())) {
+      auto &Argument = Group.value();
+      VoidConstResult VoidConst = isVoidConst(&Argument.Type());
+      if (VoidConst.IsVoid) {
+        if (FunctionType->Arguments().size() > 1) {
+          // More than 1 void argument.
+          return true;
+        }
 
-    // Filter out invalid functions.
-    auto *FunctionType = dyn_cast<CABIFunctionType>(T.get());
-    if (FunctionType) {
-      // Remove functions with more than one `void` argument.
-      for (auto &Group : llvm::enumerate(FunctionType->Arguments())) {
-        auto &Argument = Group.value();
-        VoidConstResult VoidConst = isVoidConst(&Argument.Type());
-        if (VoidConst.IsVoid) {
-          if (FunctionType->Arguments().size() > 1) {
-            // More than 1 void argument.
-            ToDrop.insert(T.get());
-            break;
-          }
-
-          if (VoidConst.IsConst) {
-            // Cannot have const void argument.
-            ToDrop.insert(T.get());
-            break;
-          }
+        if (VoidConst.IsConst) {
+          // Cannot have const void argument.
+          return true;
         }
       }
     }
   }
 
+  return false;
+}
+
+void model::fixModel(TupleTree<model::Binary> &Model) {
+  std::set<const model::Type *> ToDrop;
+
+  for (UpcastablePointer<model::Type> &T : Model->Types()) {
+    if (shouldDrop(T))
+      ToDrop.insert(T.get());
+  }
```

Comment:
Next time you can use `llvm::copy_if()` `:)`
No biggie.

Reply:
I am not the biggest fun of the paradigm (I don't think it improves readability always :) ), but I understand part of the motivation. The code would be a bit messy, since ```ToDrop``` expects ```Type *```, so we cannot do simply as:
```
  llvm::copy_if(Model->Types(),
                std::back_inserter(ToDrop),
                [](UpcastablePointer<model::Type> &T) {
                  return shouldDrop(T);
                });
```